addons/website_sale_delivery/controllers/main.py

```diff
@@ -292,3 +340,43 @@ def _update_website_sale_delivery_return(self, order, **post):
                 'new_amount_total_raw': order.amount_total,
             }
         return {}
+
+
+class WebsiteSaleDeliveryPayment(PaymentPortal):
+    @http.route(
+        '/shop/payment/transaction/<int:order_id>', type='json', auth='public', website=True
+    )
+    def shop_payment_transaction(self, order_id, access_token, **kwargs):
+
+        order = request.website.sale_get_order()
+        if not order and 'sale_last_order_id' in request.session:
+            last_order_id = request.session['sale_last_order_id']
+            order = request.env['sale.order'].sudo().browse(last_order_id).exists()
+        order_location = json.loads(order.access_point_address) if order.access_point_address else False
+
+        if not order_location:
+            return super().shop_payment_transaction(order_id, access_token, **kwargs)
+
+        # retreive all the data :
+        # name, street, city, state, zip, country
+        name = order_location['pick_up_point_name']
+        street = order_location['pick_up_point_address']
+        city = order_location['pick_up_point_town']
+        zip_code = order_location['pick_up_point_postal_code']
+        country = request.env['res.country'].search([('code', '=', order_location['pick_up_point_country'])]).id
+        state = request.env['res.country.state'].search(['&', ('code', '=', order_location['pick_up_point_state']), ('country_id.id', '=', country)]).id if (order_location['pick_up_point_state'] and country)else None
+        parent_id = order.partner_shipping_id.id
+
+        # we can check if the current partner has a partner of type "delivery" that has the same address
+        existing_partner = request.env['res.partner'].search(['&', '&', '&', ('street', '=', street), ('city', '=', city), ('state_id', '=', state), ('country_id', '=', country)])
+
+        if existing_partner:
+            if len(existing_partner) > 1:
+                order.partner_shipping_id = existing_partner[0]
+            else:
+                order.partner_shipping_id = existing_partner
+        else:
+            # if not, we create that res.partner
+            order.partner_shipping_id = request.env['res.partner'].create({'parent_id': parent_id, 'type': 'delivery', 'name': name, 'street': street, 'city': city, 'state_id': state, 'zip': zip_code, 'country_id': country})
+
+        return super().shop_payment_transaction(order_id, access_token, **kwargs)
```

Comment:
I would do it in the action_confirm of sale_order

Reply:
It would be more flexible because if the route is not used for a reason (e.g. custom code or another flow) then we will miss it